### PR TITLE
Add ethereum pair with keccak_256 hash

### PR DESCRIPTION
**Pull request by fewensa**
Some chains use Ethereum account style. and use `keccak_256` hash algorithm, like Moonbeam and Darwinia 2.0
  https://github.com/PureStake/moonbeam/blob/214154dbb724ec6c9c97a25d60eb3e74307a45c5/primitives/account/src/lib.rs#L164-L165
  The ecdsa pair is use `blake2_256` hash algorithm. Some clients, such as subxt, will not have a suitable signature algorithm to sign. now the polkadot.js is support this feature https://github.com/polkadot-js/common/blob/7611ec8b014780499a5a1d3b89cf76077d478dea/packages/keyring/src/pair/index.ts#L40

Related:
https://github.com/polkadot-js/common/issues/506
https://github.com/polkadot-js/common/pull/589

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2195668

File: primitives/core/src/ethereum.rs [hunk 1]
```diff
@@ -0,0 +1,869 @@
...
+		// This is safe due to the limited range of possible valid ids.
+		r.0[64] = recid.to_i32() as u8;
+		r
+	}
+}
+
+/// Derive a single hard junction.
+#[cfg(feature = "full_crypto")]
+fn derive_hard_junction(secret_seed: &Seed, cc: &[u8; 32]) -> Seed {
+	("Secp256k1HDKD", secret_seed, cc).using_encoded(sp_core_hashing::blake2_256)
```
Comment (fewensa) [inline primitives/core/src/ethereum.rs:367]: I'm not sure here. keep using `blake2_256`?

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2197365

**Comment by davxy** [comment]
The difference between `ecdsa.rs` and `ethereum.rs` is really few lines, i.e. only the hashing algorithm (blake2 end keccak, respectively). I'm not sure this huge code verbatim duplication is the right way to eventually introduce this.

Even the naming of these two objects looks wrong to me, these in the end are just `ecdsa+secp256k1+hash` where has is `keccak` or `blake`.

If we really need to introduce this new pair type in our codebase, a good approach IMO is to make the code in  `ecdsa.rs` generic over the hash (and maybe in the future over the curve as well... but now is not necessary). Then have two specializations (in the same source file), one using keccak and one using blake2. And obviously create two separate `CryptoTypeId` as well.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2197459

**Review by davxy — CHANGES_REQUESTED** [review]
The difference between the code in `ecdsa.rs` and `ethereum.rs` is really few lines, i.e. only the hashing algorithm (blake2 end keccak, respectively). I'm not sure this huge code verbatim duplication is the right way to eventually introduce this.

Even the naming of these two objects looks wrong to me, these in the end are just ecdsa+secp256k1+hash where has is keccak or blake.

If we really need to introduce this new key pair type directly in our codebase, a better approach IMO is to make the code in `ecdsa.rs` generic over the hash (and maybe in the future over the curve as well... but now is not necessary). Then have two specializations (in the same source file), one using keccak and one using blake2. And obviously create two separate `CryptoTypeId` as well.

**Comment by fewensa** [comment]
> The difference between the code in `ecdsa.rs` and `ethereum.rs` is really few lines, i.e. only the hashing algorithm (blake2 end keccak, respectively). I'm not sure this huge code verbatim duplication is the right way to eventually introduce this.
> 
> Even the naming of these two objects looks wrong to me, these in the end are just ecdsa+secp256k1+hash where has is keccak or blake.
> 
> If we really need to introduce this new key pair type directly in our codebase, a better approach IMO is to make the code in `ecdsa.rs` generic over the hash (and maybe in the future over the curve as well... but now is not necessary). Then have two specializations (in the same source file), one using keccak and one using blake2. And obviously create two separate `CryptoTypeId` as well.

Yes, agree with you. I can refactor it if the idea is feasible.

**Comment by bkchr** [comment]
Sounds like a good idea by @davxy

**Comment by fewensa** [comment]
Hi @davxy , I have a question there.

I have tried to add a generic type to `Signature` and Pair to support different hashing. but the `Signature` struct has `PassByInner` macro. 
https://github.com/fewensa/substrate/commit/09c5afd4b55a19f7515475fd383db0edb9cf8630#diff-a1370f23ed063d050c75648f95fc2811b04b9c04203376f24c2f7e99b00b800bR209-R212

https://github.com/paritytech/substrate/blob/ea387c634715793f806286abf1e64cabf9b7026f/primitives/runtime-interface/proc-macro/src/pass_by/inner.rs#L98

I think we can not remove `if unnamed.unnamed.len() == 1` in `PassByInner` macro.
If not use the generic method, here’s how to add a transitive hashing algorithm. otherwise, only add different struct here to make the difference between hashing algorithms

**Comment by bkchr** [comment]
@fewensa you can also implement the trait manually.

**Pull request closed without merging**